### ISSUE TITLE
Instance with __getitem__ should be iterable

### ISSUE THREAD

**Issue by Jacco** [skulpt/skulpt#469]
```python
class iterable:

    def __getitem__(self, i):
        if i < 13:
            return i
        else:
            raise IndexError("index out of range")

x = iterable()
    
for v in x:
    print v
```

Skulpt gives error: class iterable is not iterable

**Comment by rixner** [comment]
I don't think you even need ```__len__```.  It is sufficient to just have ```__getitem__``` to be iterable.

**Comment by Jacco** [comment]
Hm, indeed, didn't test that. Updated the issue to make it simpler :-)

**Comment by bnmnetp** [comment]
Sadly, the error is thrown here where we are looking for an internal definition of ``tp$iter``

```javascript
Sk.abstr.iter = function (obj) {
    if (obj.tp$iter) {
        return obj.tp$iter();
    }
    else {
        throw new Sk.builtin.TypeError("'" + Sk.abstr.typeName(obj) + "' object is not iterable");
    }
};
```

**Comment by bnmnetp** [comment]
Interestingly, I always thought that for an object to be iterable that it had to conform to the iterable protocol and provide an ``__iter__`` and ``next`` method.

Going a little deeper, I have a fix, but its a hack.  It seems like maybe the compiler out to generate different code...

```
        klass.prototype.tp$iter = function () {
            var ret;
            var iterf = this.tp$getattr("__iter__");
            var tname = Sk.abstr.typeName(this);
            if (iterf) {
                ret = Sk.misceval.callsim(iterf);
                return ret;
            }
            var gitem = this.tp$getattr("__getitem__");
            if (gitem) {
                this.iteridx = 0
                return this;
            }
            throw new Sk.builtin.TypeError("'" + tname + "' object is not iterable");
        };
        klass.prototype.tp$iternext = function () {
            var iternextf = this.tp$getattr("next");
            if (iternextf) {
                return Sk.misceval.callsim(iternextf);
            }
            var gitem = this.tp$getattr("__getitem__");
            if (gitem) {
                print(arguments[0])
                ret = Sk.misceval.callsim(gitem,this.iteridx);
                this.iteridx++;
                return ret;
            }
        };
```

**Comment by waywaaard** [comment]
CPython impl. uses a check a check first:

```c
PyObject *
PyObject_GetIter(PyObject *o)
{
    PyTypeObject *t = o->ob_type;
    getiterfunc f = NULL;
    f = t->tp_iter;
    if (f == NULL) {
        if (PySequence_Check(o))
            return PySeqIter_New(o);
        return type_error("'%.200s' object is not iterable", o);
    }
    else {
        PyObject *res = (*f)(o);
        if (res != NULL && !PyIter_Check(res)) {
            PyErr_Format(PyExc_TypeError,
                         "iter() returned non-iterator "
                         "of type '%.100s'",
                         res->ob_type->tp_name);
            Py_DECREF(res);
            res = NULL;
        }
        return res;
    }
}
```
```PySequence_Check``` returns true all sequences and checks for ```sq_item```

```c
int
PySequence_Check(PyObject *s)
{
    if (PyDict_Check(s))
        return 0;
    return s != NULL && s->ob_type->tp_as_sequence &&
        s->ob_type->tp_as_sequence->sq_item != NULL;
}
```

As you can see in the GetIter function a special sequence iterator is created and this checks for ```__getitem__``` (https://github.com/python/cpython/blob/efac0438f42e03f0ece269ac8f1616bc6dc2d8bd/Objects/iterobject.c#L12).

And here is the iter next function, that uses __getitem__

```c
static PyObject *
iter_iternext(PyObject *iterator)
{
    seqiterobject *it;
    PyObject *seq;
    PyObject *result;

    assert(PySeqIter_Check(iterator));
    it = (seqiterobject *)iterator;
    seq = it->it_seq;
    if (seq == NULL)
        return NULL;
    if (it->it_index == PY_SSIZE_T_MAX) {
        PyErr_SetString(PyExc_OverflowError,
                        "iter index too large");
        return NULL;
    }

    result = PySequence_GetItem(seq, it->it_index);
    if (result != NULL) {
        it->it_index++;
        return result;
    }
    if (PyErr_ExceptionMatches(PyExc_IndexError) ||
        PyErr_ExceptionMatches(PyExc_StopIteration))
    {
        PyErr_Clear();
        Py_DECREF(seq);
        it->it_seq = NULL;
    }
    return NULL;
}
```

I hope that helps.

**Comment by bnmnetp** [comment]
I was thinking that the solution should involve some kind of generated object, and a check at the beginning of the for loop iteration.  But I had not found the Python implementation of each.

This would not actually be to hard to put together a proper fix now.

**Comment by bnmnetp** [comment]
fixed by #469
